extensions/variance/src/main/java/io/druid/query/aggregation/variance/VarianceAggregatorFactory.java

```diff
@@ -0,0 +1,252 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.variance;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.google.common.base.Preconditions;
+import com.google.common.primitives.Doubles;
+import com.google.common.primitives.Longs;
+import com.metamx.common.StringUtils;
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.AggregatorFactoryNotMergeableException;
+import io.druid.query.aggregation.Aggregators;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.column.ValueType;
+import org.apache.commons.codec.binary.Base64;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ */
+@JsonTypeName("variance")
+public class VarianceAggregatorFactory extends AggregatorFactory
+{
+  private static final byte CACHE_TYPE_ID = 0x4;
+
+  private final String fieldName;
+  private final String name;
+
+  @JsonCreator
+  public VarianceAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName
+  )
+  {
+    Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+
+    this.name = name;
+    this.fieldName = fieldName;
+  }
+
+  @Override
+  public String getTypeName()
+  {
+    return "variance";
+  }
+
+  @Override
+  public int getMaxIntermediateSize()
+  {
+    return Longs.BYTES + Doubles.BYTES + Doubles.BYTES;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    ValueType valueType = metricFactory.columnType(fieldName);
+    if (valueType == null) {
+      return Aggregators.noopAggregator();
+    }
+    if (valueType == ValueType.FLOAT) {
+      return new VarianceAggregator.FloatInput(
+          name,
+          metricFactory.makeFloatColumnSelector(fieldName)
+      );
+    }
+    if (valueType == ValueType.LONG) {
+      return new VarianceAggregator.LongInput(
+          name,
+          metricFactory.makeLongColumnSelector(fieldName)
+      );
+    }
+    return new VarianceAggregator.ObjectInput(
+        name,
+        metricFactory.makeObjectColumnSelector(fieldName)
+    );
+  }
+
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    ValueType valueType = metricFactory.columnType(fieldName);
```

Comment:
is this aggregator supposed to be used at query time or both query and indexing time. 
can you add a unit test that does indexing and querying, both using this aggregator and see if valueType you get is appropriate. Most of the time you are going to get valueType to be COMPLEX or null except only when this aggregator is used at query time only (and data was indexed using a long/double type aggregator) at the segment scan time (rest of the processing will see valueType to be COMPLEX only)

can't you just use ObjectInput Variance aggregator and change that code in aggregate(..) to be

```
Object obj = selector.get();
if (obj != null) {
if (obj instance of VarianceHolder) {
VarianceHolder.combineValues(holder, obj);
} else {
//use instanceof to check for type to be Number, String and convert into a double
}
}
}
```
that would be more general too.